File: src/api/func/render/geometry.py
# render/geometry.py — la geometria de la ESCENA: zonas poligonales.
#
# Este modulo es EL SEGUNDO CAJON de StreamSession, y esa division es la decision de
# fondo del trabajo de zonas (spec 2026-08-28 §3). Lo que ya vivia en la sesion —
# tracker, suavizador, trazas— se olvida cuando cambia el modelo, porque todo eso
# depende de tracker_id y unas identidades nacidas con otro pipeline no significan
# nada. La geometria es al reves: describe LA ESCENA, no el modelo. Comparar dos
# modelos sobre la misma zona es exactamente el caso de uso del banco de pruebas, asi
# que borrar el poligono al cambiar de modelo seria destruir el experimento.
#
# De ahi la regla: sync() vacia el cajon del tracker y NO toca este.
#
# Lo que si mata la geometria es el cierre de la conexion, y eso es a proposito: la
# zona vale para la escena que se esta mirando, y cambiar de fuente cierra el WS. NO
# se persiste en ningun lado (decidido con el usuario el 2026-08-28): persistirla
# obligaria a identificar la fuente con una clave, y el nombre de archivo es fragil
# —"Prueba.mp4" y "Prueba_4x3.mp4" son escenas distintas, y renombrar rompe todo—.
#
# Las coordenadas se guardan NORMALIZADAS en [0,1], fracciones del ancho y del alto.
# El motivo no es portabilidad entre resoluciones (sin persistencia no hay nada que
# portar): es que asi el CLIENTE no necesita saber la resolucion del frame. Su editor
# dibuja sobre un <svg viewBox="0 0 1 1"> y convierte el click con
# (clientX - rect.left) / rect.width — sin canvas.width en ninguna parte. El backend,
# que ya calcula h, w para el auto_scale, multiplica una vez al construir la zona.
#
# stateful=false NO apaga esto. El camino one-shot de imagenes declara que no hay
# memoria TEMPORAL que construir; una zona sobre una foto suelta es perfectamente
# legitima (contar vehiculos en una region de una imagen). Por eso este cajon no se
# gatea por _stateful, y el del tracker si.

import logging
from dataclasses import dataclass
from typing import Optional, Sequence, Tuple

import numpy as np
import supervision as sv

from .annotators import annotators_for
from .draw_config import DrawConfig, anclaje_de_zona

logger = logging.getLogger(__name__)

# Topes defensivos. No son limites de producto: son la frontera entre "el usuario
# dibujo algo" y "algo esta mandando basura por el canal de control".
MAX_ZONAS = 8
MIN_VERTICES = 3
MAX_VERTICES = 64
# Area minima en coordenadas normalizadas. 1e-6 es una millonesima del frame: por
# debajo de eso el poligono es una raya (vertices colineales) y sv.PolygonZone
# construiria una mascara que nunca dispara.
AREA_MINIMA = 1e-6
# Cuanto puede diferir el aspecto declarado por el cliente del aspecto real del frame
# antes de que la zona se considere dibujada sobre otra escena (ver _aspecto_coincide).
TOLERANCIA_ASPECTO = 0.01


class GeometriaInvalida(ValueError):
    """
    Geometria que no se puede aceptar, con un motivo legible para el ack.

    Se levanta en el PARSEO, nunca en el hot path: el frame prefiere dibujar algo
    antes que caerse, mismo criterio que un box_style desconocido.
    """


@dataclass(frozen=True)
class Zona:
    """Un poligono de la escena, en coordenadas normalizadas [0,1]."""
    id: str
    puntos: np.ndarray          # (N,2) float32, x e y como fracciones del frame


@dataclass(frozen=True)
class ZonaViva:
    """Una zona ya materializada en pixeles para una resolucion concreta."""
    zona: Zona
    poligono: sv.PolygonZone
    annotator: sv.PolygonZoneAnnotator


def _punto_valido(p) -> Tuple[float, float]:
    """Un par [x, y] finito, recortado a [0,1]."""
    if not isinstance(p, (list, tuple)) or len(p) != 2:
        raise GeometriaInvalida("cada punto tiene que ser un par [x, y]")
    try:
        x, y = float(p[0]), float(p[1])
    except (TypeError, ValueError):
        raise GeometriaInvalida("las coordenadas tienen que ser numeros")
    if not (np.isfinite(x) and np.isfinite(y)):
        raise GeometriaInvalida("las coordenadas tienen que ser finitas")
    # Se RECORTA en vez de rechazar: arrastrar un vertice mas alla del borde del
    # canvas es un gesto legitimo ("que la zona llegue hasta el borde"), y el cliente
    # no tiene por que impedirlo con aritmetica propia.
    return (min(max(x, 0.0), 1.0), min(max(y, 0.0), 1.0))


def _area_normalizada(puntos: np.ndarray) -> float:
    """Area del poligono por la formula del cordon de zapato (shoelace), sin signo."""
    x, y = puntos[:, 0], puntos[:, 1]
    return float(abs(np.dot(x, np.roll(y, -1)) - np.dot(y, np.roll(x, -1))) / 2.0)


def parsear_geometria(payload: dict):
    """
    Valida el mensaje de control del cliente y devuelve (frame_wh, [Zona, ...]).

    El mensaje es DECLARATIVO Y COMPLETO, no incremental: trae toda la geometria
    vigente, no un delta. Un delta obliga a las dos puntas a coincidir sobre un
    historial, y ese es justo el acoplamiento que este proyecto viene evitando.

    Por eso mismo el rechazo es del MENSAJE ENTERO y no por zona: aceptar la mitad
    dejaria al backend con un estado que el cliente no cree tener, que es peor que
    rechazar. Ante error se conserva la geometria anterior y el ack lo explica.
    """
    if not isinstance(payload, dict):
        raise GeometriaInvalida("el mensaje de control tiene que ser un objeto")

    frame = payload.get("frame")
    if not isinstance(frame, dict):
        raise GeometriaInvalida("falta 'frame' con el tamano sobre el que se dibujo")
    try:
        fw, fh = int(frame["w"]), int(frame["h"])
    except (KeyError, TypeError, ValueError):
        raise GeometriaInvalida("'frame' tiene que traer 'w' y 'h' enteros")
    if fw <= 0 or fh <= 0:
        raise GeometriaInvalida("el tamano del frame tiene que ser positivo")

    # Las lineas de conteo todavia no existen (van en la tanda siguiente, reusando
    # este mismo canal con dos vertices). Se rechaza explicito en vez de ignorarlas
    # en silencio: un cliente que las mande y no vea nada no tendria como enterarse.
    lineas = payload.get("lines")
    if lineas:
        raise GeometriaInvalida("las lineas de conteo todavia no estan implementadas")

    crudas = payload.get("zones")
    if crudas is None:
        crudas = []
    if not isinstance(crudas, list):
        raise GeometriaInvalida("'zones' tiene que ser una lista")
    if len(crudas) > MAX_ZONAS:
        raise GeometriaInvalida("como mucho %d zonas a la vez" % MAX_ZONAS)

    zonas = []
    vistos = set()
    for cruda in crudas:
        if not isinstance(cruda, dict):
            raise GeometriaInvalida("cada zona tiene que ser un objeto")
        zid = cruda.get("id")
        if not isinstance(zid, str) or not zid or len(zid) > 32:
            raise GeometriaInvalida("cada zona necesita un 'id' de 1 a 32 caracteres")
        if zid in vistos:
            raise GeometriaInvalida("id de zona repetido: %s" % zid)
        vistos.add(zid)

        puntos = cruda.get("points")
        if not isinstance(puntos, list):
            raise GeometriaInvalida("la zona %s no trae 'points'" % zid)
        if not (MIN_VERTICES <= len(puntos) <= MAX_VERTICES):
            raise GeometriaInvalida(
                "la zona %s tiene %d vertices: van de %d a %d"
                % (zid, len(puntos), MIN_VERTICES, MAX_VERTICES))

        arr = np.array([_punto_valido(p) for p in puntos], dtype=np.float32)
        if _area_normalizada(arr) < AREA_MINIMA:
            # Vertices colineales o encimados: el poligono no encierra nada y su
            # mascara nunca dispararia. Decirlo es mas util que dibujar una raya.
            raise GeometriaInvalida("la zona %s no encierra area" % zid)
        zonas.append(Zona(id=zid, puntos=arr))

    return (fw, fh), zonas


def _aspecto_coincide(frame_wh, resolution_wh) -> bool:
    """
    Si la zona se dibujo sobre un frame de la misma forma que el que esta llegando.

    Es una ASERCION, no una funcion: sin persistencia esto no puede dispararse en uso
    normal —la fuente tiene la resolucion que tiene y la zona muere con ella—, asi que
    si se dispara es un bug. Por eso loguea y no dibuja, y no tiene interfaz: hacerle
    UI seria inflar el panel por un caso que no ocurre.

    Se guardan w y h y no el cociente porque son estrictamente mas informacion: el
    mensaje puede decir "se dibujo sobre 1440x1080 y estan llegando frames de
    1920x1080" en vez de "1.333 != 1.778".

    Tolerancia 1%: los aspectos reales estan lejisimos entre si (4:3 = 1,333 ·
    3:2 = 1,500 · 16:10 = 1,600 · 16:9 = 1,778; el par mas cercano se lleva 4%) y el
    ruido de redondeo es despreciable (854x480 da 1,779 contra 1,778, o sea 0,08%).
    """
    aw, ah = frame_wh
    bw, bh = resolution_wh
    if ah <= 0 or bh <= 0:
        return False
    a, b = aw / ah, bw / bh
    return abs(a - b) <= TOLERANCIA_ASPECTO * max(a, b)


class SceneGeometry:
    """
    El cajon de la ESCENA de una conexion: las zonas y el tamano sobre el que se
    dibujaron. Sobrevive al cambio de modelo; muere con la conexion.

    No es thread-safe y no necesita serlo: el protocolo del stream es de un frame en
    vuelo y el canal de control comparte ese mismo bucle, asi que nunca hay dos
    llamadas simultaneas.
    """

    def __init__(self):
        self._zonas: Tuple[Zona, ...] = ()
        self._frame_wh: Optional[Tuple[int, int]] = None
        # Version de la geometria. Es parte de la clave del cache de zonas vivas, con
        # el mismo criterio (y por la misma razon) que la version de DrawConfig: que
        # el hot path no reconstruya nada mientras nadie toque nada.
        self._version = 0
        self._cache_clave = None
        self._cache_zonas: Tuple[ZonaViva, ...] = ()
        # Para no repetir el aviso de aspecto una vez por frame.
        self._aviso_aspecto = None

    def __len__(self) -> int:
        return len(self._zonas)

    @property
    def vacia(self) -> bool:
        return not self._zonas

    @property
    def zonas(self) -> Tuple[Zona, ...]:
        return self._zonas

    @property
    def frame_wh(self) -> Optional[Tuple[int, int]]:
        """Sobre que tamano de frame se dibujo la geometria vigente."""
        return self._frame_wh

    def set(self, frame_wh, zonas: Sequence[Zona]) -> None:
        """Reemplaza TODA la geometria (el mensaje de control es completo, no un delta)."""
        self._zonas = tuple(zonas)
        self._frame_wh = (int(frame_wh[0]), int(frame_wh[1]))
        self._version += 1
        self._cache_clave = None
        self._cache_zonas = ()
        self._aviso_aspecto = None

    def limpiar(self) -> None:
        """Borra la geometria. NO la llama sync(): solo el cierre de la conexion."""
        self.set((1, 1), ())
        self._frame_wh = None

    def zonas_vivas(self, cfg: DrawConfig, resolution_wh) -> Tuple[ZonaViva, ...]:
        """
        Las zonas materializadas en pixeles para esta resolucion y esta config.

        Se cachean contra (version de la geometria, version de la config, resolucion):
        construir un PolygonZone rasteriza una mascara del tamano del poligono, que no
        es algo para hacer por frame. La version de la config entra porque el color, el
        grosor y el ANCLAJE salen de ahi, y el anclaje cambia que se cuenta.
        """
        if not self._zonas:
            return ()

        if self._frame_wh is not None and not _aspecto_coincide(self._frame_wh, resolution_wh):
            if self._aviso_aspecto != resolution_wh:
                logger.warning(
                    "Zonas ignoradas: se dibujaron sobre %dx%d y estan llegando frames "
                    "de %dx%d. La geometria describe otra escena.",
                    self._frame_wh[0], self._frame_wh[1],
                    resolution_wh[0], resolution_wh[1])
                self._aviso_aspecto = resolution_wh
            return ()

        clave = (self._version, cfg.version, tuple(resolution_wh))
        if self._cache_clave == clave:
            return self._cache_zonas

        w, h = resolution_wh
        ann = annotators_for(cfg, (w, h))
        color = sv.Color.from_hex(cfg.zone_color)
        anclaje = anclaje_de_zona(cfg.zone_anchor)
        vivas = []
        for zona in self._zonas:
            # np.round y no int(): truncar corre el poligono medio pixel hacia el
            # origen, y sobre un frame chico eso se ve.
            pts = np.round(zona.puntos * np.array([w, h], dtype=np.float32)).astype(np.int64)
            poligono = sv.PolygonZone(polygon=pts, triggering_anchors=(anclaje,))
            vivas.append(ZonaViva(
                zona=zona,
                poligono=poligono,
                annotator=sv.PolygonZoneAnnotator(
                    zone=poligono,
                    color=color,
                    thickness=ann.thickness,
                    text_color=sv.Color.from_hex(cfg.label_color),
                    text_scale=ann.text_scale,
                    text_thickness=max(1, ann.thickness - 1),
                ),
            ))

        self._cache_clave = clave
        self._cache_zonas = tuple(vivas)
        return self._cache_zonas

File: client/src/features/inference/services/geometry.ts
// geometry.ts — la geometria de la escena del lado del cliente (zonas poligonales).
//
// DOS COSAS QUE PARECEN UNA Y NO LO SON, y separarlas es la decision de fondo:
//
//   - El COLOR de la zona y el ANCLAJE son ajustes del usuario: viven en
//     drawSettings, persisten en localStorage y viajan por POST /config/draw.
//   - El POLIGONO describe LA ESCENA: vive en streamStore, muere cuando cambia la
//     fuente y NO se persiste. Persistirlo obligaria a identificar la fuente con una
//     clave, y el nombre de archivo es fragil ("Prueba.mp4" y "Prueba_4x3.mp4" son
//     escenas distintas, y renombrar cualquiera rompe la asociacion).
//
// COORDENADAS NORMALIZADAS [0,1], fracciones del ancho y del alto. El motivo NO es
// portabilidad entre resoluciones (sin persistencia no hay nada que portar): es que
// asi ESTE lado no necesita saber la resolucion del frame. La conversion de un click
// colapsa a (clientX - rect.left) / rect.width, sin canvas.width en ninguna parte.
// (El editor SI multiplica al DIBUJAR, por px de la caja: ver el encabezado de
// ZoneEditor.tsx para por que el viewBox de 1x1 que proponia el spec no servia.)
//
// EL CANAL: el poligono viaja por el propio WebSocket, no por HTTP. La zona pertenece
// a UNA conexion (muere con ella) y un POST no sabe a que WebSocket le esta hablando;
// ademas, por el mismo canal no hay carrera con los frames en vuelo — el orden ES el
// orden. Ver render/geometry.py y el spec 2026-08-28 §5.

/** Un punto en coordenadas normalizadas: [x, y], ambos en [0,1]. */
export type Punto = [number, number];

/** Un poligono de la escena. El id lo genera el cliente y el backend lo devuelve tal cual. */
export interface Zona {
  id: string;
  puntos: Punto[];
}

/** Mensaje de control que el cliente manda por el WS. */
export interface MensajeGeometria {
  type: 'geometry';
  /** Sobre que tamano de frame se dibujo (§4.1 del spec): dos enteros, no el cociente. */
  frame: { w: number; h: number };
  zones: { id: string; points: Punto[] }[];
}

/**
 * Mensaje de control que pone en cero el acumulado de una zona (o de todas, sin `id`).
 *
 * Va por el mismo canal que la geometria y por la misma razon: el contador vive en ESTA
 * conexion. A diferencia del mensaje de geometria, este NO se recuerda para re-emitirlo
 * al reconectar — una conexion nueva ya nace con la cuenta en cero.
 */
export interface MensajeResetZona {
  type: 'zone_reset';
  id?: string | null;
}

/**
 * Empezar o terminar el volcado de detecciones a disco.
 *
 * Va por el WS y no por HTTP por lo mismo que la geometria: lo que se exporta son las
 * detecciones de ESTA conexion. Bajar el archivo, en cambio, si va por HTTP — puede
 * pesar megabytes y no tiene por que competir con los frames.
 */
export interface MensajeExport {
  type: 'export_start' | 'export_stop';
}

/** Cualquier mensaje de control que el cliente manda por el WS. */
export type MensajeControl = MensajeGeometria | MensajeResetZona | MensajeExport;


// ── Los acks ──────────────────────────────────────────────────────────────────
// Union DISCRIMINADA por `type` y no un objeto laxo: por el canal de control ya viajan
// tres respuestas distintas, y el dia que se agregue una cuarta el compilador tiene que
// obligar a atenderla en vez de dejarla pasar como `unknown`.

/** `error` != null significa que se conservo la geometria anterior. */
export interface AckGeometria {
  type: 'geometry_ack';
  zones: number;
  lines: number;
  error: string | null;
}

export interface AckResetZona {
  type: 'zone_reset_ack';
  zone: string | null;
  error: string | null;
}

/** Estado efectivo del volcado de detecciones de esta conexion. */
export interface AckExport {
  type: 'export_ack';
  recording: boolean;
  file?: string | null;
  rows?: number;
  frames?: number;
  error?: string | null;
}

/** Un control que el backend no supo interpretar. */
export interface AckControlDesconocido {
  type: 'control_error';
  error: string;
}

export type Ack = AckGeometria | AckResetZona | AckExport | AckControlDesconocido;

/** Minimo de vertices para que un poligono encierre algo (coincide con el backend). */
export const MIN_VERTICES = 3;
/** Tope de zonas simultaneas (coincide con MAX_ZONAS de render/geometry.py). */
export const MAX_ZONAS = 8;

/**
 * El mensaje declarativo COMPLETO: siempre va toda la geometria vigente, nunca un
 * delta. Un delta obligaria a las dos puntas a coincidir sobre un historial, que es
 * justo el acoplamiento que este proyecto viene evitando en el resto de la superficie.
 */
export function mensajeGeometria(frameW: number, frameH: number, zonas: Zona[]): MensajeGeometria {
  return {
    type: 'geometry',
    frame: { w: Math.round(frameW), h: Math.round(frameH) },
    zones: zonas.map((z) => ({ id: z.id, points: z.puntos })),
  };
}

/** Si el mensaje que llego por el WS es la respuesta a un control y no a un frame. */
export function esAck(mensaje: unknown): mensaje is Ack {
  // El envelope de un frame es {task, result, error}: no tiene 'type'. El discriminador
  // es el simetrico del que usa el backend (_decode_control en mainAPI.py).
  return (
    typeof mensaje === 'object' &&
    mensaje !== null &&
    typeof (mensaje as { type?: unknown }).type === 'string'
  );
}

/** Id corto y unico dentro de la escena. No sale del cliente mas que en el mensaje. */
export function nuevoIdDeZona(existentes: Zona[]): string {
  let n = existentes.length + 1;
  const usados = new Set(existentes.map((z) => z.id));
  while (usados.has(`z${n}`)) n += 1;
  return `z${n}`;
}

/** Area del poligono (formula del cordon de zapato), para descartar los degenerados. */
export function area(puntos: Punto[]): number {
  let acc = 0;
  for (let i = 0; i < puntos.length; i += 1) {
    const [x1, y1] = puntos[i];
    const [x2, y2] = puntos[(i + 1) % puntos.length];
    acc += x1 * y2 - x2 * y1;
  }
  return Math.abs(acc) / 2;
}

/**
 * Si el poligono es aceptable. El backend valida igual y es la autoridad; esto existe
 * para no ofrecerle al usuario un boton "Confirmar" que va a fallar del otro lado.
 */
export function zonaValida(puntos: Punto[]): boolean {
  return puntos.length >= MIN_VERTICES && area(puntos) > 1e-6;
}

/**
 * La caja de contenido del canvas RELATIVA a su contenedor posicionado.
 *
 * El canvas se centra con `max-h-full max-w-full` y sin width/height en CSS, asi que
 * el navegador le conserva el aspecto del bitmap y su rect ES la imagen. Pero el
 * overlayRoot es `absolute inset-0` sobre TODO el workspace, que es mas grande. Sin
 * esto el SVG del editor quedaria estirado sobre las bandas vacias y cada click
 * caeria unos pixeles corrido.
 *
 * Se deriva midiendo, no asumiendo: si algun dia el CSS fuerza width y height a la
 * vez, object-contain empezaria a poner bandas ADENTRO del elemento y la formula
 * ingenua se romperia en silencio.
 */
export function cajaDelCanvas(canvas: HTMLCanvasElement, contenedor: HTMLElement) {
  const c = canvas.getBoundingClientRect();
  const p = contenedor.getBoundingClientRect();
  const aspectoBitmap = canvas.width / canvas.height;
  const aspectoCaja = c.width / c.height;

  // Con object-contain, si los aspectos no coinciden hay bandas adentro del elemento.
  let w = c.width;
  let h = c.height;
  if (Number.isFinite(aspectoBitmap) && aspectoBitmap > 0 && Number.isFinite(aspectoCaja)) {
    if (aspectoCaja > aspectoBitmap) w = c.height * aspectoBitmap;
    else h = c.width / aspectoBitmap;
  }
  return {
    left: c.left - p.left + (c.width - w) / 2,
    top: c.top - p.top + (c.height - h) / 2,
    width: w,
    height: h,
  };
}
